tests/unit/transport/test_zeromq.py

```diff
@@ -305,3 +309,221 @@ def test_send(self):
+        gather.start()
+        # Allow time for server channel to start, especially on windows
+        time.sleep(2)
+        for i in range(send_num):
+            expect.append(i)
+            load = {'tgt_type': 'glob', 'tgt': '*', 'jid': i}
+            server_channel.publish(load)
+        server_channel.publish(
+            {'tgt_type': 'glob', 'tgt': '*', 'stop': True}
+        )
+        gather.join()
+        server_channel.pub_close()
+        assert len(results) == send_num, (len(results), set(expect).difference(results))
+
+    def test_publish_to_pubserv_tcp(self):
+        '''
+        Test sending 10K messags to ZeroMQPubServerChannel using TCP transport
+        '''
+        opts = dict(self.master_config, ipc_mode='tcp', pub_hwm=0)
+        server_channel = salt.transport.zeromq.ZeroMQPubServerChannel(opts)
+        server_channel.pre_fork(self.process_manager, kwargs={
+            'log_queue': salt.log.setup.get_multiprocessing_logging_queue()
+        })
+        pub_uri = 'tcp://{interface}:{publish_port}'.format(**server_channel.opts)
+        send_num = 10000
+        expect = []
+        results = []
+        gather = threading.Thread(target=self._gather_results, args=(self.minion_config, pub_uri, results,))
+        gather.start()
+        # Allow time for server channel to start, especially on windows
+        time.sleep(2)
+        for i in range(send_num):
+            expect.append(i)
+            load = {'tgt_type': 'glob', 'tgt': '*', 'jid': i}
+            server_channel.publish(load)
+        gather.join()
+        server_channel.pub_close()
+        assert len(results) == send_num, (len(results), set(expect).difference(results))
+
+    @staticmethod
+    def _send_small(opts, sid, num=10):
+        server_channel = salt.transport.zeromq.ZeroMQPubServerChannel(opts)
+        for i in range(num):
+            load = {'tgt_type': 'glob', 'tgt': '*', 'jid': '{}-{}'.format(sid, i)}
+            server_channel.publish(load)
+
+    @staticmethod
+    def _send_large(opts, sid, num=10, size=250000 * 3):
+        server_channel = salt.transport.zeromq.ZeroMQPubServerChannel(opts)
+        for i in range(num):
+            load = {'tgt_type': 'glob', 'tgt': '*', 'jid': '{}-{}'.format(sid, i), 'xdata': '0' * size}
+            server_channel.publish(load)
+
+    def test_issue_36469_tcp(self):
+        '''
+        Test sending both large and small messags to publisher using TCP
+        '''
+        opts = dict(self.master_config, ipc_mode='tcp', pub_hwm=0)
+        server_channel = salt.transport.zeromq.ZeroMQPubServerChannel(opts)
+        server_channel.pre_fork(self.process_manager, kwargs={
+            'log_queue': salt.log.setup.get_multiprocessing_logging_queue()
+        })
+        send_num = 10 * 4
+        expect = []
+        results = []
+        pub_uri = 'tcp://{interface}:{publish_port}'.format(**opts)
+        # Allow time for server channel to start, especially on windows
+        time.sleep(2)
```

Comment:
This is not a good sign. What if needed 3 seconds? Or overloaded CI 10?.. Or 1? Or none at all? We are waiting on server channel, so we should rather look for if it is ready.

Reply:
I agree but I am not sure how to check if the daemon process is started without adding some synchronization logic. Unfortunately we can't check if the sockets are connected because zmq will always allow you to connect even if the other side is not bound or connected. I'm open to other ideas.

Reply:
btw, generally for this kind of thing I would use a `multiprocessing.Event` and have the get set after the sockets are connected up and we're starting the main daemon loop. We can check the even from the parent process to know if the daemon is 'ready'. This approach has the added benefit being able to signal the daemon to stop from the parent process. Still, If going down that path, I'd rather do it in develop once this gets merged up.

Reply:
OK I can't lie, so I still honestly disagree :rofl: to keep such tests as Unit and have them fail just because of timing that we even don't know how to test.

@cachedout can we agree that we call it "technical debt" and then @dwoz gets an extra-cycle to move these type of tests away from Unit into something like `tests/functional`, that are almost like integration, but not against full-blown minion/master?

Reply:
@isbm I'm happy to start moving things to `tests/functional` in the develop branch. @cachedout We're good here?

Reply:
@isbm That is a reasonable compromise. Let's do that.